### ISSUE TITLE
SessionStatus#setComplete() should affect concurrent requests as well [SPR-16306]

### ISSUE THREAD

**Issue by spring-issuemaster** [spring-projects/spring-framework#20853]
**[Sergey Yaskov](https://jira.spring.io/secure/ViewProfile.jspa?name=yaskovdev)** opened **[SPR-16306](https://jira.spring.io/browse/SPR-16306?redirect=false)** and commented

I have the Spring MVC controller which handles AJAX requests from a user. I have the next 2 handler methods:

```java
@PostMapping("/notifications")
void performLongRunningOperation() {
    // some operation which may take about 10 seconds
}

@PostMapping("/cancel")
void cancel(SessionStatus sessionStatus) {
    sessionStatus.setComplete();
}
```

The problem is that if `cancel()` finishes execution before `performLongRunningOperation()` finishes execution, then `setComplete()` has no effect. More precisely, `setComplete()` cleans the session up, but then the session unexpectedly gets restored to the state which was during execution of the `performLongRunningOperation()` method. This is the unexpected behavior.

To summarize: `setComplete()` method does not work properly in a situation when one do not know exactly the order in which the handler methods are going to finish execution.

Please see the link to the related StackOverflow question in the "Reference URL".

The reproduction project could be found here: https://github.com/yaskovdev/sandbox/tree/master/spring-session-sandbox

---

**Affects:** 4.3.10

**Reference URL:** https://stackoverflow.com/questions/46505612/spring-mvc-setcomplete-does-not-guarantee-the-session-cleanup

**Comment by rstoyanchev** [comment]
As mentioned by Juergen above this isn't straight forward to implement, and `SessionStatus` was really designed around a form POST workflow. The cancellation should rather probably await and/or cancel any ongoing operations, before completing the session.
